lib/firestore/delete.js

```diff
@@ -280,18 +320,119 @@ FirestoreDelete.progressBar = new ProgressBar("Deleted :current docs (:rate docs
  */
 FirestoreDelete.prototype._recursiveBatchDelete = function() {
   var self = this;
-  return this._getDescendantBatch(this.allDescendants, this.batchSize).then(function(docs) {
-    if (docs.length <= 0) {
-      return Promise.resolve();
+
+  // Tunable deletion parameters
+  var readBatchSize = 5000;
+  var deleteBatchSize = 250;
+  var maxPendingDeletes = 15;
+  var maxQueueSize = deleteBatchSize * maxPendingDeletes * 2;
+
+  // All temporary variables for the deletion queue.
+  var queue = [];
+  var numPendingDeletes = 0;
+  var pagesRemaining = true;
+  var pageIncoming = false;
+  var lastDocName;
+
+  var retried = {};
+  var failures = [];
+
+  var queueLoop = function() {
+    if (queue.length == 0 && numPendingDeletes == 0 && !pagesRemaining) {
+      return true;
     }
 
-    return self._deleteDocuments(docs).then(function(numDeleted) {
-      // Tick the progress bar
-      FirestoreDelete.progressBar.tick(numDeleted);
+    if (queue.length <= maxQueueSize && pagesRemaining && !pageIncoming) {
+      pageIncoming = true;
 
-      // Recurse to delete another batch
-      return self._recursiveBatchDelete();
-    });
+      self
+        ._getDescendantBatch(self.allDescendants, readBatchSize, lastDocName)
+        .then(function(docs) {
+          pageIncoming = false;
+
+          if (docs.length == 0) {
+            pagesRemaining = false;
+            return;
+          }
+
+          docs.forEach(function(doc) {
+            queue.push(doc);
+          });
+
+          lastDocName = docs[docs.length - 1].name;
+        })
+        .catch(function(e) {
+          logger.debug("Failed to fetch page after " + lastDocName, e);
+          pageIncoming = false;
+        });
+    }
+
+    if (numPendingDeletes > maxPendingDeletes) {
+      return false;
+    }
+
+    if (queue.length == 0) {
+      return false;
+    }
+
+    var toDelete = [];
+    var numToDelete = Math.min(deleteBatchSize, queue.length);
+
+    for (var i = 0; i < numToDelete; i++) {
+      toDelete.push(queue.shift());
+    }
+
+    numPendingDeletes++;
+    self
+      ._deleteDocuments(toDelete)
+      .then(function(numDeleted) {
+        FirestoreDelete.progressBar.tick(numDeleted);
+        numPendingDeletes--;
+      })
+      .catch(function(e) {
+        // For server errors, retry if the document has not yet been retried.
+        if (e.status >= 500 && e.status < 600 && !retried[toDelete.name]) {
+          logger.debug("Server error deleting doc " + toDelete.name, e);
```

Comment:
From line 378 it seems like toDelete is a list? But you're calling `toDelete.name` in line 394 and line 395 as if it was a single object.